awswrangler/torch.py

```diff
@@ -0,0 +1,456 @@
+"""PyTorch Module."""
+import io
+import logging
+import os
+import pathlib
+import re
+import tarfile
+from collections.abc import Iterable
+from io import BytesIO
+from typing import Any, Callable, Iterator, List, Optional, Tuple, Union
+
+import boto3  # type: ignore
+import numpy as np  # type: ignore
+import sqlalchemy  # type: ignore
+import torch  # type: ignore
+import torchaudio  # type: ignore
+from PIL import Image  # type: ignore
+from torch.utils.data.dataset import Dataset, IterableDataset  # type: ignore
+from torchvision.transforms.functional import to_tensor  # type: ignore
+
+from awswrangler import _utils, db, s3
+
+_logger: logging.Logger = logging.getLogger(__name__)
+
+
+class _BaseS3Dataset:
+    """PyTorch Amazon S3 Map-Style Dataset."""
+
+    def __init__(
+        self, path: Union[str, List[str]], suffix: Optional[str] = None, boto3_session: Optional[boto3.Session] = None
+    ):
+        """PyTorch Map-Style S3 Dataset.
+
+        Parameters
+        ----------
+        path : Union[str, List[str]]
+            S3 prefix (e.g. s3://bucket/prefix) or list of S3 objects paths (e.g. [s3://bucket/key0, s3://bucket/key1]).
+        boto3_session : boto3.Session(), optional
+            Boto3 Session. The default boto3 session will be used if boto3_session receive None.
+
+        Returns
+        -------
+        torch.utils.data.Dataset
+
+        """
+        super().__init__()
+        self._session = _utils.ensure_session(session=boto3_session)
+        self._paths: List[str] = s3._path2list(  # pylint: disable=protected-access
+            path=path, suffix=suffix, boto3_session=self._session
+        )
+
+    def _fetch_data(self, path: str) -> Any:
+        """Add parquet and csv support."""
+        bucket, key = _utils.parse_path(path=path)
+        buff = BytesIO()
+        client_s3: boto3.client = _utils.client(service_name="s3", session=self._session)
+        client_s3.download_fileobj(Bucket=bucket, Key=key, Fileobj=buff)
+        buff.seek(0)
+        return buff
+
+    @staticmethod
+    def _load_data(data: io.BytesIO, path: str) -> Any:
+        if path.endswith(".pt"):
+            data = torch.load(data)
+        elif path.endswith(".tar.gz") or path.endswith(".tgz"):
+            tarfile.open(fileobj=data)
+            raise NotImplementedError("Tar loader not implemented!")
+            # tar = tarfile.open(fileobj=data)
+            # for member in tar.getmembers():
+        else:
+            raise NotImplementedError()
+
+        return data
+
+
+class _ListS3Dataset(_BaseS3Dataset, Dataset):
+    """PyTorch Amazon S3 Map-Style List Dataset."""
+
+    def __getitem__(self, index):
+        path = self._paths[index]
+        data = self._fetch_data(path)
+        return [self._data_fn(data), self._label_fn(path)]
+
+    def __len__(self):
+        return len(self._paths)
+
+    def _data_fn(self, data) -> Any:
+        pass
+
+    def _label_fn(self, path: str) -> Any:
+        pass
```

Comment:
As `_ListS3Dataset` is apparently an abstract class, consider `raise NotImplementedError()` instead of `pass`.